Accept selection state from mutating select elements

aknuds1 opened the pull request:
Don't change incoming selection state when mutating `select` elements.

Fixes #58

aknuds1 committed:
Don't change selection state when mutating select elements

kristoferjoseph commented:
Did you run the tests after making this change?

aknuds1 commented:
@kristoferjoseph Looks as if I omitted the tests, unfortunately. Tried now and one test is failing.

aknuds1 commented:
@kristoferjoseph Specifically, the test 'input values get copied' fails. Why is it that the `value` of the mutated element should be kept, as opposed to be updated to the `value` of the mutating element?

kristoferjoseph commented:
OK.
SELECT may be a special case.
Do me a favor and add a test for the SELECT behavior you are expecting.
Might need your add one more if else clause.

✌

On Sun, Jan 8, 2017, 10:55 AM Arve Knudsen <notifications@github.com> wrote:

> @kristoferjoseph <https://github.com/kristoferjoseph> Looks as if I
> omitted the tests, unfortunately. Tried now and one test is failing.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/maxogden/yo-yo/pull/59#issuecomment-271170796>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAEqFL_QBzJG8fH1sw1DYNOseFuqOhcvks5rQTEYgaJpZM4LX7XC>
> .
>

aknuds1 commented:
@kristoferjoseph The test in question doesn't really make sense to me, I think we better resolve how this is supposed to work before we go any further. To me, it seems that the test should assert that `newEl.value` gets copied to `el`. Is this wrong?

kristoferjoseph commented:
It makes sure that if _no_ value is on the _new_ element that the _old_
value is copied over.
Otherwise every DOM update would lose the existing value in your input.

On Sun, Jan 8, 2017, 11:02 AM Arve Knudsen <notifications@github.com> wrote:

> @kristoferjoseph <https://github.com/kristoferjoseph> The test in
> question doesn't really make sense to me, I think we better resolve how
> this is supposed to work before we go any further. To me, it seems that the
> test should assert that newEl.value gets copied to el. Is this wrong?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/maxogden/yo-yo/pull/59#issuecomment-271171285>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAEqFIPMgQo-Pui4eKRx3N9vDvMb-o3hks5rQTK-gaJpZM4LX7XC>
> .
>

aknuds1 commented:
@kristoferjoseph It also ensures that new values don't get copied over however, I just checked.

aknuds1 commented:
What if you want to erase an input's value when you re-render, how do you control this if yo-yo ignores this change? Doesn't seem to make sense to me.

aknuds1 commented:
The pattern that would make sense to me is to record when an input's `value` changes and render this into DOM updates.

aknuds1 commented:
@kristoferjoseph I see now that yo-yo checks if the mutating element has `value` different from null. However, this actually doesn't detect that the test sets `newEl.value` directly, so maybe there's a bit of confusion due to this.

kristoferjoseph commented:
The thinking behind this is that the input element is for user input. It
errors on the side of the input value being changed by the user.

I agree you could add tests and a case for when the new input has a value
already.

On Sun, Jan 8, 2017, 11:14 AM Arve Knudsen <notifications@github.com> wrote:

> The pattern that would make sense to me is to record when an input's value
> changes and render this into DOM updates.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/maxogden/yo-yo/pull/59#issuecomment-271172001>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAEqFC1pHocR9tYzeB7-k1VSMegeBR2rks5rQTWFgaJpZM4LX7XC>
> .
>

aknuds1 commented:
@kristoferjoseph I rewrote the original test slightly and added a new one, asserting that new input values get copied if they are actually set.

aknuds1 commented:
However, I have to deal with the select case.

aknuds1 commented:
@kristoferjoseph I tested one more thing, the case where you clear the `value` via the mutating element. I tested by assigning an empty string to `value`, and I can see that the old value is not overwritten. Does this make sense to you? To me it seems wrong.

aknuds1 commented:
I've reverted to the original behaviour and added a test for the clearing of input value case.

Will have to add handling of selects, as per the bug I experienced.

aknuds1 commented:
@kristoferjoseph Does it ever make sense to keep the `selectedIndex` (which yo-yo attempted to do via `value ) on a `select`? Can it ever be null?

Even if the select has no elements, the selectedIndex will be -1 and in that case I don't see any point in maintaining it as the mutated select will lose all its children anyway.

aknuds1 commented:
I've made the change in my branch now to not change the selection state of select elements. I don't see any case where it makes sense, although I could be missing something.

kristoferjoseph commented:
@aknuds1 as for keeping the selected index. Imagine a scenario where you have set the select index then update something else in the same scope. You want to make sure the index is maintained during an unrelated update.

aknuds1 commented:
@kristoferjoseph How would you formulate this into a test case? I can't see any way that `selectedIndex` would be null, so there's no way to tell whether it's supposed to be updated or not. That's why it becomes so tricky to try to tell whether the incoming mutation is trying to change a value or not.

aknuds1 commented:
I think it's way safer to set all properties you need on the mutating element, rather than to expect yo-yo to maintain them.

kristoferjoseph commented:
@aknuds1 https://github.com/yoshuawuyts/nanomorph/pull/28/files

aknuds1 commented:
@kristoferjoseph That doesn't show how to test this for `select` elements though, which is what I would need to know? Like I said, I cannot see any way to do it in practice.

kristoferjoseph commented:
Ah, sorry, was actually replying to one of your earlier questions. I will make a select test as well. Was hoping this patch might fix some of the other issues you were seeing.

aknuds1 commented:
@kristoferjoseph Is it necessarily a great idea to implement this sort of intelligence into nanomorph? I suspect it violates the principle of least surprise, at least I was surprised to find that yo-yo tries to preserve old values as opposed to copying directly from the incoming/mutating element.

kristoferjoseph commented:
I would say that it is more surprising when an input value is lost when a
parent node is updated.

Imagine the use case. You type something into an input then do something
unrelated, like open a drop-down and your input value is lost.

Does that make it more clear?

On Tue, Jan 10, 2017, 6:30 AM Arve Knudsen <notifications@github.com> wrote:

> @kristoferjoseph <https://github.com/kristoferjoseph> Is it necessarily a
> great idea to implement this sort of intelligence into nanomorph? I suspect
> it violates the principle of least surprise, at least I was surprised to
> find that yo-yo tries to preserve old values as opposed to copying directly
> from the incoming/mutating element.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/maxogden/yo-yo/pull/59#issuecomment-271589408>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAEqFOqY5cE9zlLQkuAHBXA7JkDs26Ouks5rQ5XrgaJpZM4LX7XC>
> .
>

aknuds1 commented:
@kristoferjoseph Yeah it makes it more clear, but I'm just not so sure a morphing library should do this. It sounds like a decision to be made on top of the morphing, i.e. it might be undesirable to the user of the morphing library. The value preservation algorithm certainly is imperfect as it stands, as my scenario with the select elements shows.

I'm discussing with @yoshuawuyts however, and he explained why it's useful in the choo context. I think it really depends on higher level concerns, so I'm just not so sure the behaviour should be hardcoded into nanomorph. I don't know how it works in morphdom, but I suspect it doesn't have this sort of intelligence built in, since yo-yo is implementing it on top of it. I suspect it would be better handled by a module around nanomorph.

yoshuawuyts commented:
> I'm discussing with @yoshuawuyts however

@kristoferjoseph come to IRC 🕥 :sparkles: :tada:

aknuds1 commented:
It's good we're having this discussion though, I wasn't aware of the issues you can run into without preserving values until @yoshuawuyts explained it to me.

kristoferjoseph commented:
Happy to hear it is becoming more clear. This is one concept that is
definitely different in use vs theory. Which makes it a contentious issue.

I'm terrible at IRC or I'd be there already 😬

On Tue, Jan 10, 2017, 7:17 AM Arve Knudsen <notifications@github.com> wrote:

> It's good we're having this discussion though, I wasn't aware of the
> issues you can run into without preserving values until @yoshuawuyts
> <https://github.com/yoshuawuyts> explained it to me.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/maxogden/yo-yo/pull/59#issuecomment-271601788>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAEqFG9awZK1eS94ip5X5RhmguGn6Yc3ks5rQ6DtgaJpZM4LX7XC>
> .
>

yoshuawuyts commented:
@kristoferjoseph http://irccloud.com/ and then join freenode and #choo :D - _I believe in you_
